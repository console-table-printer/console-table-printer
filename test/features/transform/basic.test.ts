import { Table } from '../../../index';
import { ValueTransformer, CellValue } from '../../../src/models/external-table';

describe('Testing transform functionality', () => {
  it('should allow adding columns with transformers', () => {
    const numberFormatter: ValueTransformer = (data: CellValue): CellValue =>
      Number(data).toFixed(2);

    const p = new Table()
      .addColumn('name')
      .addColumn({ name: 'salary', transformer: numberFormatter })
      .addRow({ name: 'John', salary: 50000 })
      .addRow({ name: 'Jane', salary: 60000.5 });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should allow adding multiple columns with different transformers', () => {
    const currencyFormatter: ValueTransformer = (data: CellValue): CellValue =>
      `$${Number(data).toFixed(2)}`;
    
    const uppercaseFormatter: ValueTransformer = (data: CellValue): CellValue =>
      String(data).toUpperCase();

    const p = new Table()
      .addColumn({ name: 'name', transformer: uppercaseFormatter })
      .addColumn({ name: 'salary', transformer: currencyFormatter })
      .addColumn('department')
      .addRow({ name: 'john doe', salary: 50000, department: 'Engineering' })
      .addRow({ name: 'jane smith', salary: 75000.5, department: 'Marketing' });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle null and undefined values in transformers', () => {
    const safeFormatter: ValueTransformer = (data: CellValue): CellValue => {
      if (data === null || data === undefined) {
        return 'N/A';
      }
      return `Formatted: ${data}`;
    };

    const p = new Table()
      .addColumn('id')
      .addColumn({ name: 'value', transformer: safeFormatter })
      .addRow({ id: 1, value: 'test' })
      .addRow({ id: 2, value: null })
      .addRow({ id: 3, value: undefined })
      .addRow({ id: 4 }); // missing value property

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle string transformers', () => {
    const stringFormatter: ValueTransformer = (data: CellValue): CellValue =>
      String(data).replace(/\s+/g, '_').toLowerCase();

    const p = new Table()
      .addColumn('original')
      .addColumn({ name: 'transformed', transformer: stringFormatter })
      .addRow({ original: 'Hello World', transformed: 'Hello World' })
      .addRow({ original: 'Test String', transformed: 'Test String' })
      .addRow({ original: 'Multiple   Spaces', transformed: 'Multiple   Spaces' });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle number transformers with different formats', () => {
    const percentageFormatter: ValueTransformer = (data: CellValue): CellValue =>
      `${(Number(data) * 100).toFixed(1)}%`;

    const scientificFormatter: ValueTransformer = (data: CellValue): CellValue =>
      Number(data).toExponential(2);

    const p = new Table()
      .addColumn('decimal')
      .addColumn({ name: 'percentage', transformer: percentageFormatter })
      .addColumn({ name: 'scientific', transformer: scientificFormatter })
      .addRow({ decimal: 0.25, percentage: 0.25, scientific: 0.25 })
      .addRow({ decimal: 0.1234, percentage: 0.1234, scientific: 0.1234 })
      .addRow({ decimal: 1.5, percentage: 1.5, scientific: 1.5 });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should work with chained column additions', () => {
    const formatter1: ValueTransformer = (data: CellValue): CellValue =>
      `[${data}]`;
    
    const formatter2: ValueTransformer = (data: CellValue): CellValue =>
      `{${data}}`;

    const p = new Table()
      .addColumn({ name: 'col1', transformer: formatter1 })
      .addColumn('col2')
      .addColumn({ name: 'col3', transformer: formatter2 })
      .addRow({ col1: 'A', col2: 'B', col3: 'C' })
      .addRow({ col1: 'X', col2: 'Y', col3: 'Z' });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle transformers with complex column configurations', () => {
    const complexFormatter: ValueTransformer = (data: CellValue): CellValue =>
      `*** ${String(data).toUpperCase()} ***`;

    const p = new Table()
      .addColumn({
        name: 'complexColumn',
        title: 'Complex Transformed Column',
        alignment: 'center',
        color: 'blue',
        transformer: complexFormatter,
      })
      .addColumn('simpleColumn')
      .addRow({ complexColumn: 'hello', simpleColumn: 'world' })
      .addRow({ complexColumn: 'test', simpleColumn: 'data' });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle transformers that return different data types', () => {
    const mixedFormatter: ValueTransformer = (data: CellValue): CellValue => {
      if (typeof data === 'number') {
        return data * 2;
      }
      if (typeof data === 'string') {
        return data.length;
      }
      return 'unknown';
    };

    const p = new Table()
      .addColumn('input')
      .addColumn({ name: 'transformed', transformer: mixedFormatter })
      .addRow({ input: 'hello', transformed: 'hello' })
      .addRow({ input: 42, transformed: 42 })
      .addRow({ input: 'world', transformed: 'world' })
      .addRow({ input: 100, transformed: 100 });

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });

  it('should handle mathematical transformers correctly across multiple renders', () => {
    // Test with a mathematical transformer to ensure no double-transformation
    const cubeTransformer: ValueTransformer = (data: CellValue): CellValue => {
      const num = Number(data);
      return num * num * num; // Cube the number
    };

    const p = new Table()
      .addColumn('number')
      .addColumn({ name: 'cubed', transformer: cubeTransformer })
      .addRow({ number: 2, cubed: 2 })   // 2³ = 8
      .addRow({ number: 3, cubed: 3 })   // 3³ = 27
      .addRow({ number: 4, cubed: 4 });  // 4³ = 64

    // Render multiple times to ensure consistency
    const render1 = p.render();
    const render2 = p.render();
    const render3 = p.render();

    // All renders should be identical
    expect(render1).toEqual(render2);
    expect(render2).toEqual(render3);

    // Verify the output contains the correct cubed values
    expect(render1).toContain('8');   // 2³ = 8
    expect(render1).toContain('27');  // 3³ = 27  
    expect(render1).toContain('64');  // 4³ = 64

    // Verify it doesn't contain values that would result from double-cubing
    expect(render1).not.toContain('512');   // 2³³ = 512
    expect(render1).not.toContain('19683'); // 3³³ = 19683
    expect(render1).not.toContain('262144'); // 4³³ = 262144

    p.printTable();
    expect(p.render()).toMatchSnapshot();
  });
});
