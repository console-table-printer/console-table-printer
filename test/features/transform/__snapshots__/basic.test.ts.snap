// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Testing transform functionality should allow adding columns with transformers 1`] = `
"┌──────┬──────────┐
│[37m [0m[01mname[0m │[37m [0m[01m  salary[0m │
├──────┼──────────┤
│[37m [0m[37mJohn[0m │[37m [0m[37m50000.00[0m │
│[37m [0m[37mJane[0m │[37m [0m[37m60000.50[0m │
└──────┴──────────┘"
`;

exports[`Testing transform functionality should allow adding multiple columns with different transformers 1`] = `
"┌────────────┬───────────┬─────────────┐
│[37m [0m[01m      name[0m │[37m [0m[01m   salary[0m │[37m [0m[01m department[0m │
├────────────┼───────────┼─────────────┤
│[37m [0m[37m  JOHN DOE[0m │[37m [0m[37m$50000.00[0m │[37m [0m[37mEngineering[0m │
│[37m [0m[37mJANE SMITH[0m │[37m [0m[37m$75000.50[0m │[37m [0m[37m  Marketing[0m │
└────────────┴───────────┴─────────────┘"
`;

exports[`Testing transform functionality should handle mathematical transformers correctly across multiple renders 1`] = `
"┌────────┬───────┐
│[37m [0m[01mnumber[0m │[37m [0m[01mcubed[0m │
├────────┼───────┤
│[37m [0m[37m     2[0m │[37m [0m[37m    8[0m │
│[37m [0m[37m     3[0m │[37m [0m[37m   27[0m │
│[37m [0m[37m     4[0m │[37m [0m[37m   64[0m │
└────────┴───────┘"
`;

exports[`Testing transform functionality should handle null and undefined values in transformers 1`] = `
"┌────┬─────────────────┐
│[37m [0m[01mid[0m │[37m [0m[01m          value[0m │
├────┼─────────────────┤
│[37m [0m[37m 1[0m │[37m [0m[37mFormatted: test[0m │
│[37m [0m[37m 2[0m │[37m [0m[37m            N/A[0m │
│[37m [0m[37m 3[0m │[37m [0m[37m            N/A[0m │
│[37m [0m[37m 4[0m │[37m [0m[37m            N/A[0m │
└────┴─────────────────┘"
`;

exports[`Testing transform functionality should handle number transformers with different formats 1`] = `
"┌─────────┬────────────┬────────────┐
│[37m [0m[01mdecimal[0m │[37m [0m[01mpercentage[0m │[37m [0m[01mscientific[0m │
├─────────┼────────────┼────────────┤
│[37m [0m[37m   0.25[0m │[37m [0m[37m     25.0%[0m │[37m [0m[37m   2.50e-1[0m │
│[37m [0m[37m 0.1234[0m │[37m [0m[37m     12.3%[0m │[37m [0m[37m   1.23e-1[0m │
│[37m [0m[37m    1.5[0m │[37m [0m[37m    150.0%[0m │[37m [0m[37m   1.50e+0[0m │
└─────────┴────────────┴────────────┘"
`;

exports[`Testing transform functionality should handle string transformers 1`] = `
"┌───────────────────┬─────────────────┐
│[37m [0m[01m         original[0m │[37m [0m[01m    transformed[0m │
├───────────────────┼─────────────────┤
│[37m [0m[37m      Hello World[0m │[37m [0m[37m    hello_world[0m │
│[37m [0m[37m      Test String[0m │[37m [0m[37m    test_string[0m │
│[37m [0m[37mMultiple   Spaces[0m │[37m [0m[37mmultiple_spaces[0m │
└───────────────────┴─────────────────┘"
`;

exports[`Testing transform functionality should handle transformers that return different data types 1`] = `
"┌───────┬─────────────┐
│[37m [0m[01minput[0m │[37m [0m[01mtransformed[0m │
├───────┼─────────────┤
│[37m [0m[37mhello[0m │[37m [0m[37m          5[0m │
│[37m [0m[37m   42[0m │[37m [0m[37m         84[0m │
│[37m [0m[37mworld[0m │[37m [0m[37m          5[0m │
│[37m [0m[37m  100[0m │[37m [0m[37m        200[0m │
└───────┴─────────────┘"
`;

exports[`Testing transform functionality should handle transformers with complex column configurations 1`] = `
"┌────────────────────────────┬──────────────┐
│[37m [0m[01mComplex Transformed Column[0m │[37m [0m[01msimpleColumn[0m │
├────────────────────────────┼──────────────┤
│[37m [0m[34m      *** HELLO ***       [0m │[37m [0m[37m       world[0m │
│[37m [0m[34m       *** TEST ***       [0m │[37m [0m[37m        data[0m │
└────────────────────────────┴──────────────┘"
`;

exports[`Testing transform functionality should work with chained column additions 1`] = `
"┌──────┬──────┬──────┐
│[37m [0m[01mcol1[0m │[37m [0m[01mcol2[0m │[37m [0m[01mcol3[0m │
├──────┼──────┼──────┤
│[37m [0m[37m [A][0m │[37m [0m[37m   B[0m │[37m [0m[37m {C}[0m │
│[37m [0m[37m [X][0m │[37m [0m[37m   Y[0m │[37m [0m[37m {Z}[0m │
└──────┴──────┴──────┘"
`;
